exifread/classes.py

```diff
@@ -69,73 +186,77 @@ def __init__(self, file_handle: BinaryIO, endian, offset, fake_exif, strict: boo
-        # Construct a format string from the requested length and signedness;
-        # raise a ValueError if length is something silly like 3
-        try:
-            fmt += {
-                (1, False): 'B',
-                (1, True):  'b',
-                (2, False): 'H',
-                (2, True):  'h',
-                (4, False): 'I',
-                (4, True):  'i',
-                (8, False): 'L',
-                (8, True):  'l',
-                }[(length, signed)]
-        except KeyError as err:
-            raise ValueError('unexpected unpacking length: %d' % length) from err
-        self.file_handle.seek(self.offset + offset)
-        buf = self.file_handle.read(length)
-
-        if buf:
-            # https://github.com/ianare/exif-py/pull/158
-            # had to revert as this certain fields to be empty
-            # please provide test images
-            return struct.unpack(fmt, buf)[0]
-        return 0
+    @cached_property
+    def ifd_attrs(self):
+        return IfdAttrs(self.magic_number)
+
+    @cached_property
+    def tag_attrs(self):
+        return TagAttrs(self.magic_number)
+
+    def offset_to_next_ifd(self, ifd, entries):
+        return ifd + self.ifd_attrs.ntag_length + entries * self.tag_attrs.total_length
 
     def n2b(self, offset, length) -> bytes:
         """Convert offset to bytes."""
-        s = b''
+        s = b""
         for _ in range(length):
-            if self.endian == 'I':
+            if self.endian == "I":
                 s += bytes([offset & 0xFF])
             else:
                 s = bytes([offset & 0xFF]) + s
             offset = offset >> 8
         return s
 
     def _first_ifd(self) -> int:
-        """Return first IFD."""
-        return self.s2n(4, 4)
+        """Return the pointer to first IFD."""
+        return s2n(
+            self.file_handle,
+            offset=self.offset + self.ifd_attrs.offset_to_first_ifd,
+            length=self.ifd_attrs.offset_to_ifd_length,
+            signed=False,
+            endian=self.endian,
+        )
 
     def _next_ifd(self, ifd) -> int:
         """Return the pointer to next IFD."""
-        entries = self.s2n(ifd, 2)
-        next_ifd = self.s2n(ifd + 2 + 12 * entries, 4)
+        entries = s2n(
+            file_handle=self.file_handle,
+            offset=self.offset + ifd,
+            length=self.ifd_attrs.ntag_length,
+            signed=False,
+            endian=self.endian,
+        )
+        next_ifd = s2n(
+            file_handle=self.file_handle,
+            offset=self.offset + self.offset_to_next_ifd(ifd, entries),
+            length=self.ifd_attrs.offset_to_ifd_length,
+            endian=self.endian,
+        )
+        # bytesize of offsets is 8 in bigtiff and 4 in tiff
         if next_ifd == ifd:
             return 0
         return next_ifd
 
     def list_ifd(self) -> list:
         """Return the list of IFDs in the header."""
         i = self._first_ifd()
+        # i = 16
```

Comment:
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement.  -->
**nitpick:** ‏Code comment